diff --git a/bsp/at32/libraries/rt_drivers/config/f402_405/dma_config.h b/bsp/at32/libraries/rt_drivers/config/f402_405/dma_config.h
index c379be4185..d095aa302a 100644
--- a/bsp/at32/libraries/rt_drivers/config/f402_405/dma_config.h
+++ b/bsp/at32/libraries/rt_drivers/config/f402_405/dma_config.h
@@ -179,6 +179,13 @@ extern "C" {
 #define UART4_TX_DMA_IRQ                DMA2_Channel1_IRQn
 #define UART4_TX_DMA_MUX_CHANNEL        DMA2MUX_CHANNEL1
 #define UART4_TX_DMA_REQ_ID             DMAMUX_DMAREQ_ID_USART4_TX
+#elif defined(BSP_UART2_RX_USING_DMA) && !defined(UART2_RX_DMA_CHANNEL)
+#define UART2_RX_DMA_IRQHandler         DMA2_Channel1_IRQHandler
+#define UART2_RX_DMA_CLOCK              CRM_DMA2_PERIPH_CLOCK
+#define UART2_RX_DMA_CHANNEL            DMA2_CHANNEL1
+#define UART2_RX_DMA_IRQ                DMA2_Channel1_IRQn
+#define UART2_RX_DMA_MUX_CHANNEL        DMA2MUX_CHANNEL1
+#define UART2_RX_DMA_REQ_ID             DMAMUX_DMAREQ_ID_USART2_RX
 #endif
 
 /* DMA2 channel2 */
@@ -189,6 +196,13 @@ extern "C" {
 #define UART5_RX_DMA_IRQ                DMA2_Channel2_IRQn
 #define UART5_RX_DMA_MUX_CHANNEL        DMA2MUX_CHANNEL2
 #define UART5_RX_DMA_REQ_ID             DMAMUX_DMAREQ_ID_USART5_RX
+#elif defined(BSP_UART2_TX_USING_DMA) && !defined(UART2_TX_DMA_CHANNEL)
+#define UART2_TX_DMA_IRQHandler         DMA2_Channel2_IRQHandler
+#define UART2_TX_DMA_CLOCK              CRM_DMA2_PERIPH_CLOCK
+#define UART2_TX_DMA_CHANNEL            DMA2_CHANNEL2
+#define UART2_TX_DMA_IRQ                DMA2_Channel2_IRQn
+#define UART2_TX_DMA_MUX_CHANNEL        DMA2MUX_CHANNEL2
+#define UART2_TX_DMA_REQ_ID             DMAMUX_DMAREQ_ID_USART2_TX
 #endif
 
 /* DMA2 channel3 */
