commit 1531c17f1d95045566fb4ea2005ed4ce674de9db
Author: Koen De Vleeschauwer <kdv@kdvelectronics.eu>
Date:   Wed Aug 21 16:42:08 2024 +0200

    [bsp/at32] static dma_isr() in hard_i2c and spi

diff --git a/bsp/at32/libraries/rt_drivers/drv_hard_i2c.c b/bsp/at32/libraries/rt_drivers/drv_hard_i2c.c
index b6f5202e5..ad62ce866 100644
--- a/bsp/at32/libraries/rt_drivers/drv_hard_i2c.c
+++ b/bsp/at32/libraries/rt_drivers/drv_hard_i2c.c
@@ -1337,7 +1337,7 @@ void i2c_err_isr(struct at32_i2c_handle *handle)
     i2c_interrupt_enable(handle->i2c_x, I2C_ERR_INT, FALSE);
 }
 
-void dma_isr(struct at32_i2c_handle *handle)
+static void dma_isr(struct at32_i2c_handle *handle)
 {
     volatile rt_uint32_t reg_sts = 0, index = 0;
     struct dma_config *dma = RT_NULL;
diff --git a/bsp/at32/libraries/rt_drivers/drv_spi.c b/bsp/at32/libraries/rt_drivers/drv_spi.c
index 16125ddd4..a127ac811 100644
--- a/bsp/at32/libraries/rt_drivers/drv_spi.c
+++ b/bsp/at32/libraries/rt_drivers/drv_spi.c
@@ -571,7 +571,7 @@ static void at32_spi_dma_init(struct at32_spi *instance)
     }
 }
 
-void dma_isr(struct dma_config *dma_instance)
+static void dma_isr(struct dma_config *dma_instance)
 {
     volatile rt_uint32_t reg_sts = 0, index = 0;
 
